File: src/main/java/de/xclipsen/ircbridge/AuctionHouseUnderbidFeature.java
package de.xclipsen.ircbridge;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import de.xclipsen.ircbridge.mixin.HandledScreenAccessor;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.ingame.HandledScreen;
import net.minecraft.client.option.KeyBinding;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.NbtComponent;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.tooltip.TooltipType;
import net.minecraft.screen.GenericContainerScreenHandler;
import net.minecraft.screen.slot.Slot;
import net.minecraft.text.HoverEvent;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import org.slf4j.Logger;

import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class AuctionHouseUnderbidFeature {
	private static final Gson GSON = new Gson();
	private static final Pattern AUCTION_PRICE_PATTERN = Pattern.compile("(?:Buy it now|Starting bid|Top bid):\\s*([0-9,]+)\\s+coins", Pattern.CASE_INSENSITIVE);
	private static final Pattern ALLOWED_SCREEN_PATTERN = Pattern.compile("Auctions Browser|Manage Auctions|Auctions: \".*\"?", Pattern.CASE_INSENSITIVE);
	private static final Pattern PET_LEVEL_PATTERN = Pattern.compile("\\[Lvl\\s+(\\d+)]", Pattern.CASE_INSENSITIVE);
	private static final int CREATE_BIN_ITEM_SLOT = 13;

	private final Logger logger;
	private final KeyBinding copyUnderbidKey;
	private final LowestBinService lowestBinService;
	private String lastCreateBinItemKey = "";

	public AuctionHouseUnderbidFeature(Logger logger, KeyBinding copyUnderbidKey) {
		this.logger = logger;
		this.copyUnderbidKey = copyUnderbidKey;
		this.lowestBinService = new LowestBinService(logger);
	}

	public void handleEndTick(MinecraftClient client, BridgeConfig config) {
		lowestBinService.refreshIfNeeded();
		handleCreateBinAutoCopy(client, config);

		while (copyUnderbidKey.wasPressed()) {
			copyHoveredUnderbid(client, config);
		}
	}

	public void stop() {
		lowestBinService.stop();
	}

	private void copyHoveredUnderbid(MinecraftClient client, BridgeConfig config) {
		if (client == null || client.player == null || client.currentScreen == null) {
			return;
		}

		if (!config.auctionHouseUnderbidEnabled) {
			showMessage(client, Text.literal("[AH] Underbid copy is disabled in config.").formatted(Formatting.RED));
			return;
		}

		if (!(client.currentScreen instanceof HandledScreen<?> handledScreen)) {
			return;
		}

		String title = handledScreen.getTitle().getString();
		if (!ALLOWED_SCREEN_PATTERN.matcher(title).matches()) {
			return;
		}

		Slot hoveredSlot = ((HandledScreenAccessor) handledScreen).xclipsenIrcBridge$getFocusedSlot();
		if (hoveredSlot == null || !hoveredSlot.hasStack()) {
			showMessage(client, Text.literal("[AH] Hover an auction item first.").formatted(Formatting.RED));
			return;
		}

		Long price = extractAuctionPrice(client, hoveredSlot.getStack());
		if (price == null || price <= 0L) {
			showMessage(client, Text.literal("[AH] No auction price found on the hovered item.").formatted(Formatting.RED));
			return;
		}

		long underbidPrice = Math.max(0L, price - 1L);
		client.keyboard.setClipboard(Long.toString(underbidPrice));
		showMessage(client, createSuccessMessage(underbidPrice));
	}

	private void handleCreateBinAutoCopy(MinecraftClient client, BridgeConfig config) {
		if (client == null || client.player == null || client.currentScreen == null) {
			lastCreateBinItemKey = "";
			return;
		}

		if (!config.auctionHouseCreateBinAutoCopyEnabled) {
			lastCreateBinItemKey = "";
			return;
		}

		if (!(client.currentScreen instanceof HandledScreen<?> handledScreen)) {
			lastCreateBinItemKey = "";
			return;
		}

		if (!"Create BIN Auction".equals(handledScreen.getTitle().getString())) {
			lastCreateBinItemKey = "";
			return;
		}

		if (!(handledScreen.getScreenHandler() instanceof GenericContainerScreenHandler genericHandler)) {
			return;
		}

		if (genericHandler.getInventory().size() <= CREATE_BIN_ITEM_SLOT) {
			return;
		}

		ItemStack stack = genericHandler.getInventory().getStack(CREATE_BIN_ITEM_SLOT);
		if (stack.isEmpty()) {
			lastCreateBinItemKey = "";
			return;
		}

		String internalName = extractInternalName(stack);
		if (internalName.isBlank()) {
			return;
		}

		String dedupeKey = internalName + "#" + stack.getCount();
		if (dedupeKey.equals(lastCreateBinItemKey)) {
			return;
		}

		long lowestBin = resolveLowestBin(stack, internalName);
		if (lowestBin <= 0L) {
			return;
		}

		lastCreateBinItemKey = dedupeKey;
		long underbidPrice = Math.max(0L, lowestBin * stack.getCount() - 1L);
		client.keyboard.setClipboard(Long.toString(underbidPrice));
		showMessage(client, Text.literal("")
			.append(Text.literal("[AH] ").formatted(Formatting.GREEN))
			.append(Text.literal("Auto copied Lowest BIN - 1: "))
			.append(Text.literal(formatNumber(underbidPrice)).formatted(Formatting.GOLD)));
	}

	private static Long extractAuctionPrice(MinecraftClient client, ItemStack stack) {
		List<Text> tooltip = stack.getTooltip(Item.TooltipContext.create(client.world), client.player, TooltipType.BASIC);
		for (Text line : tooltip) {
			Matcher matcher = AUCTION_PRICE_PATTERN.matcher(line.getString());
			if (!matcher.find()) {
				continue;
			}

			String digits = matcher.group(1).replace(",", "");
			try {
				return Long.parseLong(digits);
			} catch (NumberFormatException ignored) {
				return null;
			}
		}

		return null;
	}

	private long resolveLowestBin(ItemStack stack, String internalName) {
		long direct = lowestBinService.getLowestBin(internalName);
		if (direct > 0L) {
			return direct;
		}

		String leveledPetName = extractLeveledPetInternalName(stack, internalName);
		if (!leveledPetName.equals(internalName)) {
			long petPrice = lowestBinService.getLowestBin(leveledPetName);
			if (petPrice > 0L) {
				return petPrice;
			}
		}

		return -1L;
	}

	private static String extractInternalName(ItemStack stack) {
		NbtComponent customData = stack.get(DataComponentTypes.CUSTOM_DATA);
		if (customData == null) {
			return "";
		}

		var customNbt = customData.copyNbt();
		String itemId = customNbt.getString("id", "");
		if (itemId.isBlank()) {
			return "";
		}

		if (!"PET".equalsIgnoreCase(itemId)) {
			return itemId.toUpperCase(Locale.ROOT).replace(':', '-');
		}

		String petInfo = customNbt.getString("petInfo", "");
		if (petInfo.isBlank()) {
			return "PET";
		}

		try {
			JsonObject petInfoObject = GSON.fromJson(petInfo, JsonObject.class);
			if (petInfoObject == null) {
				return "PET";
			}

			String petType = petInfoObject.has("type") ? petInfoObject.get("type").getAsString() : "";
			String petTier = petInfoObject.has("tier") ? petInfoObject.get("tier").getAsString() : "";
			int rarityIndex = switch (petTier.toUpperCase(Locale.ROOT)) {
				case "COMMON" -> 0;
				case "UNCOMMON" -> 1;
				case "RARE" -> 2;
				case "EPIC" -> 3;
				case "LEGENDARY" -> 4;
				case "MYTHIC" -> 5;
				case "DIVINE" -> 6;
				default -> -1;
			};

			if (petType.isBlank() || rarityIndex < 0) {
				return "PET";
			}

			return petType.toUpperCase(Locale.ROOT) + ";" + rarityIndex;
		} catch (Exception exception) {
			return "PET";
		}
	}

	private static String extractLeveledPetInternalName(ItemStack stack, String internalName) {
		if (!internalName.contains(";")) {
			return internalName;
		}

		Matcher matcher = PET_LEVEL_PATTERN.matcher(stack.getName().getString());
		if (!matcher.find()) {
			return internalName;
		}

		return internalName + "+" + matcher.group(1);
	}

	private static MutableText createSuccessMessage(long underbidPrice) {
		return Text.literal("")
			.append(Text.literal("[AH] ").formatted(Formatting.GREEN))
			.append(Text.literal("Copied "))
			.append(Text.literal(formatNumber(underbidPrice)).formatted(Formatting.GOLD))
			.append(Text.literal(" to clipboard."));
	}

	private static void showMessage(MinecraftClient client, Text message) {
		if (client.player != null) {
			client.player.sendMessage(message, false);
		}
	}

	private static String formatNumber(long value) {
		return String.format(java.util.Locale.US, "%,d", value);
	}
}

File: src/main/kotlin/de/xclipsen/ircbridge/AuctionHouseUnderbidFeature.kt
package de.xclipsen.ircbridge

import com.google.gson.JsonParser
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.inventory.ContainerScreen
import net.minecraft.core.component.DataComponents
import net.minecraft.world.item.ItemStack
import net.minecraft.network.chat.Component
import java.util.Locale
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import kotlin.jvm.optionals.getOrNull
import kotlin.math.roundToLong

object AuctionHouseUnderbidFeature {
	private const val CREATE_BIN_AUCTION_TITLE = "Create BIN Auction"
	private const val CREATE_BIN_ITEM_SLOT = 13
	private const val PRICE_REFRESH_INTERVAL_MS = 5L * 60L * 1000L
	private const val MAX_LOCAL_STALE_MS = 60L * 60L * 1000L
	private const val FAILURE_RETRY_MS = 30_000L
	private const val MAX_CACHE_ENTRIES = 128
	private val PET_RARITIES = listOf("COMMON", "UNCOMMON", "RARE", "EPIC", "LEGENDARY", "MYTHIC")
	private val priceExecutor = Executors.newSingleThreadExecutor { runnable ->
		Thread(runnable, "xclipsen-auction-prices").apply { isDaemon = true }
	}

	private var lastCopiedFingerprint: String? = null
	@Volatile private var priceRefreshInFlight = false
	@Volatile private var lastPriceRefreshAt = 0L
	private val retryAfterByItem = mutableMapOf<String, Long>()
	@Volatile private var refreshingItemId: String? = null
	@Volatile private var lastRefreshPartial = false
	@Volatile private var lastRefreshStale = false
	@Volatile private var lastRefreshError = ""
	private val priceCache = linkedMapOf<String, CachedPrice>()
	@Volatile private var cacheOrigin = ""
	@Volatile private var refreshFuture: CompletableFuture<*>? = null

	fun onWorldChange() {
		lastCopiedFingerprint = null
		priceRefreshInFlight = false
		refreshFuture?.cancel(true)
		refreshFuture = null
		refreshingItemId = null
	}

	fun shutdown() {
		onWorldChange()
		priceExecutor.shutdownNow()
	}

	fun statusLine(): String {
		val config = XclipsenIrcBridgeClient.instance?.config() ?: return "Unavailable"
		return "enabled=${config.auctionHouseModuleEnabled && config.auctionHouseAutoCopyUnderbidEnabled}, " +
			"refreshing=$priceRefreshInFlight, cachedItems=${priceCache.size}, partial=$lastRefreshPartial, " +
			"stale=$lastRefreshStale, lastRefreshAt=$lastPriceRefreshAt, lastError=${lastRefreshError.ifBlank { "none" }}"
	}

	fun onTick(client: Minecraft) {
		val config = XclipsenIrcBridgeClient.instance?.config() ?: return
		val origin = activeModBackendBaseUrl(config)
		if (origin != cacheOrigin) {
			cacheOrigin = origin
			synchronized(priceCache) { priceCache.clear() }
			retryAfterByItem.clear()
			refreshFuture?.cancel(true)
			refreshingItemId = null
			priceRefreshInFlight = false
		}
		if (!config.auctionHouseModuleEnabled || !config.auctionHouseAutoCopyUnderbidEnabled) {
			lastCopiedFingerprint = null
			return
		}

		val screen = client.screen as? ContainerScreen ?: run {
			lastCopiedFingerprint = null
			return
		}
		if (screen.title.string != CREATE_BIN_AUCTION_TITLE) {
			lastCopiedFingerprint = null
			return
		}

		val stack = screen.menu.slots.getOrNull(CREATE_BIN_ITEM_SLOT)?.item ?: return
		if (stack.isEmpty) {
			lastCopiedFingerprint = null
			return
		}

		val internalName = resolveInternalName(stack) ?: return
		maybeRefreshPrice(internalName)
		val cached = synchronized(priceCache) { priceCache[internalName] }
			?.takeIf { System.currentTimeMillis() - it.sourceFetchedAt <= MAX_LOCAL_STALE_MS }
		val unitPrice = (cached?.lowestBin ?: cached?.bazaarSellReference)
			?.roundToLong()
			?.takeIf { it > 0L } ?: return
		if (unitPrice > Long.MAX_VALUE / stack.count.toLong()) return
		val totalPrice = unitPrice * stack.count.toLong()
		if (totalPrice <= 1L) {
			return
		}

		val fingerprint = "$internalName:${stack.count}:$unitPrice"
		if (lastCopiedFingerprint == fingerprint) {
			return
		}

		val underbidPrice = totalPrice - 1L
		client.keyboardHandler.setClipboard(underbidPrice.toString())
		client.player?.sendSystemMessage(Component.literal("Copied ${formatCoins(underbidPrice)} to clipboard. (Auto Underbid)"))
		lastCopiedFingerprint = fingerprint
	}

	private fun maybeRefreshPrice(itemId: String) {
		if (priceRefreshInFlight) {
			return
		}

		val now = System.currentTimeMillis()
		val cached = synchronized(priceCache) { priceCache[itemId] }
		if (cached != null && now - cached.receivedAt <= PRICE_REFRESH_INTERVAL_MS) {
			return
		}
		if (now < (retryAfterByItem[itemId] ?: 0L)) {
			return
		}

		priceRefreshInFlight = true
		refreshingItemId = itemId
		val generation = ClientSessionLifecycle.snapshot()
		val requestOrigin = cacheOrigin
		val backend = XclipsenIrcBridgeClient.instance?.backendBridge() ?: run {
			priceRefreshInFlight = false
			return
		}
		refreshFuture = CompletableFuture.supplyAsync({ backend.fetchAuctionHousePrices(listOf(itemId)) }, priceExecutor)
			.whenComplete { payload, throwable ->
				val client = Minecraft.getInstance()
				client.execute {
					if (!ClientSessionLifecycle.isCurrent(generation) || refreshingItemId != itemId || cacheOrigin != requestOrigin) return@execute
					priceRefreshInFlight = false
					refreshingItemId = null
					val item = payload?.items?.singleOrNull()
					if (throwable != null || item == null || (item.lowestBin == null && item.bazaarSellReference == null)) {
						retryAfterByItem[itemId] = System.currentTimeMillis() + FAILURE_RETRY_MS
						lastRefreshError = "price data unavailable"
						return@execute
					}
					val sourceTimestamp = (if (item.lowestBin != null) payload.sources.lowestBin.fetchedAt else payload.sources.bazaar.fetchedAt)
						?: return@execute
					val cachedPrice = CachedPrice(item.lowestBin, item.bazaarSellReference, sourceTimestamp, System.currentTimeMillis())
					synchronized(priceCache) {
						priceCache[itemId] = cachedPrice
						while (priceCache.size > MAX_CACHE_ENTRIES) priceCache.remove(priceCache.keys.first())
					}
					lastPriceRefreshAt = cachedPrice.receivedAt
					lastRefreshPartial = payload.partial
					lastRefreshStale = payload.sources.lowestBin.stale || payload.sources.bazaar.stale
					lastRefreshError = ""
					retryAfterByItem.remove(itemId)
				}
			}
	}

	private fun resolveInternalName(stack: ItemStack): String? {
		val customData = stack.get(DataComponents.CUSTOM_DATA) ?: return null
		val nbt = customData.copyTag()
		val rawId = nbt.getString("id").getOrNull()?.takeUnless { it.isBlank() } ?: return null
		if (rawId != "PET") {
			return rawId
		}

		val petInfoRaw = nbt.getString("petInfo").getOrNull()?.takeUnless { it.isBlank() } ?: return null
		return runCatching {
			val petInfo = JsonParser.parseString(petInfoRaw).asJsonObject
			val petType = petInfo.get("type")?.asString?.uppercase() ?: return null
			val petTier = petInfo.get("tier")?.asString?.uppercase() ?: return null
			val rarityIndex = PET_RARITIES.indexOf(petTier)
			if (rarityIndex < 0) return null
			val baseId = "$petType;$rarityIndex"
			val level = PET_LEVEL_PATTERN.find(stack.hoverName.string)?.groupValues?.get(1)?.replace(",", "")?.toIntOrNull()
			when {
				level == 200 -> "$baseId+200"
				level == 100 -> "$baseId+100"
				petType == "GOLDEN_DRAGON" && level != null && level >= 100 -> "$baseId+100"
				else -> baseId
			}
		}.getOrNull()
	}

	private fun formatCoins(value: Long): String = String.format(Locale.US, "%,d", value)

	private data class CachedPrice(
		val lowestBin: Double?,
		val bazaarSellReference: Double?,
		val sourceFetchedAt: Long,
		val receivedAt: Long,
	)

	private val PET_LEVEL_PATTERN = Regex("\\[Lvl ([0-9,]+)]")
}
